### PR TITLE
Rework parallel transform scan and reduce pattern (sycl backend)

### PR DESCRIPTION
**Pull request by reble**
Adding future implementation to dpcpp backend.
Rework parallel transform scan and reduce pattern by adding true async behavior at lowest level to sycl backend and introducing synchronization points at hetero backend level.

Succeeds #188. Feedback had been addressed.

### CODE REVIEW

**Comment by capatober** [comment]
@reble, do you have any news regarding CI failures?
At least clang-format is failed

**Comment by reble** [comment]
> @reble, do you have any news regarding CI failures?
> At least clang-format is failed

CI is all green now. Tests for ranges and FPGA backend are passing with new API.

File: include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h [hunk 1]
```diff
@@ -58,13 +58,14 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _RandomAccessIterator1 __f
...
+    auto __res = oneapi::dpl::__par_backend_hetero::__parallel_transform_reduce<_RepackedTp>(
+                     ::std::forward<_ExecutionPolicy>(__exec),
+                     unseq_backend::transform_init<_Policy, _BinaryOperation1, _Functor>{
+                         __binary_op1, _Functor{__binary_op2}},                                    // transform
+                     __binary_op1,                                                                 // combine
+                     unseq_backend::reduce<_Policy, _BinaryOperation1, _RepackedTp>{__binary_op1}, // reduce
+                     __buf1.all_view(), __buf2.all_view())
+                     .get();
 
     return __binary_op1(__init, _Tp{__res});
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h:70]: There is a host call of `__binary_op1`.  I have a couple of questions:
1) @capatober, Why that computation is not executed on a device?
2) @reble , what are you going to do with it from an async call perspective?

File: include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h [hunk 2]
```diff
@@ -89,13 +90,14 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
...
+    auto __res = oneapi::dpl::__par_backend_hetero::__parallel_transform_reduce<_RepackedTp>(
+                     ::std::forward<_ExecutionPolicy>(__exec),
+                     unseq_backend::transform_init<_Policy, _BinaryOperation, _Functor>{
+                         __binary_op, _Functor{__unary_op}},                                     // transform
+                     __binary_op,                                                                // combine
+                     unseq_backend::reduce<_Policy, _BinaryOperation, _RepackedTp>{__binary_op}, // reduce
+                     __buf.all_view())
+                     .get();
 
     return __binary_op(__init, _Tp{__res});
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h:102]: The same questions (see above)

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 3]
```diff
@@ -398,9 +397,30 @@ class __future_base
...
+        : __par_backend_hetero::__future_base(__e), __data(__b), __result_idx(__o)
+    {
+        if (sizeof...(_Ts) != 0)
+            __tmps = ::std::unique_ptr<__lifetime_keeper<_Ts...>>(new __lifetime_keeper<_Ts...>(__t...));
+    }
+
+    _T
+    get()
+    {
+        this->wait();
```
Comment (capatober) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:419]: Seems we don't need it because the next line is a synchronization point

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 4]
```diff
@@ -398,9 +397,30 @@ class __future_base
...
+        return __data.template get_access<access_mode::read>()[__result_idx];
+    }
+    template <class _Tp, class _Enable>
+    friend class oneapi::dpl::__internal::__future;
```
Comment (capatober) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:423]: Let's move this to another patch if it's not needed for this one

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 5]
```diff
@@ -398,9 +397,30 @@ class __future_base
...
+template <typename _T>
 class __future : public __future_base
 {
+    ::std::size_t __result_idx;
+    sycl::buffer<_T> __data;
+    ::std::unique_ptr<__par_backend_hetero::__lifetime_keeper_base> __tmps;
+
+  public:
+    template <typename... _Ts>
+    __future(sycl::event __e, size_t __o, sycl::buffer<_T> __b, _Ts... __t)
```
Comment (capatober) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:409]: Let's remove `__t` if corresponding cases are not used in the patch

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 6]
```diff
@@ -381,11 +381,10 @@ struct __lifetime_keeper : public __lifetime_keeper_base
 };
 
 // TODO: towards higher abstraction and generic future. implementation specific sycl::event should be hidden
-class __future_base
+struct __future_base
```
Comment (capatober) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:384]: Let's move this to another patch if it's not needed for this one

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 7]
```diff
@@ -398,9 +397,30 @@ class __future_base
...
+    {
+        if (sizeof...(_Ts) != 0)
+            __tmps = ::std::unique_ptr<__lifetime_keeper<_Ts...>>(new __lifetime_keeper<_Ts...>(__t...));
+    }
+
+    _T
+    get()
+    {
+        this->wait();
+        return __data.template get_access<access_mode::read>()[__result_idx];
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:416]: As we discussed before, the getting an access for whole __data leads to copy the __data to host....  and it is a performance gap. Creating a sub-buffer probably resolves the problem.

**Review by capatober — APPROVED** [review]
I don't see any blockers that prevent merging.
Please, wait for CI results to see if there are compilation or runtime errors.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h (hunk 7 above)
Comment (capatober) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:416]: Good point

File: include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h (hunk 1 above)
Comment (capatober) [reply include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h:70]: Here we work with host arguments: `__init` and `__res` are both on hosts.
@reble, am I right that this issue is resolved at async parallel_reduce pattern?

Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h:70]: `__first1` also my be on the host... but we pass dpc++ policy and we expect computation on a device :-)
Probably,   `__init `  should be copied to a device as well? And what if  `__init ` is on a device already? (as a result one of previous operations)

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h (hunk 7 above)
Comment (reble) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:416]: Since that would be an optimization that doesn't effect the async behavior and could be applied in both cases (blocking and non-blocking), we might consider a new PR.
